nat-lab/tests/utils/batching.py

```diff
@@ -0,0 +1,126 @@
+import asyncio
+import math
+import os
+import pytest
+import subprocess
+import typing
+from scapy.all import PcapReader, Packet  # type: ignore
+from typing import Callable, List, Optional
+
+
+def generate_histogram(data: list[int], buckets: int, bucket_size: int) -> List[int]:
+    if len(data) == 0:
+        return []
+
+    sorted_values = sorted(data)
+    max_val = sorted_values[-1]
+
+    if max_val >= buckets * bucket_size:
+        raise ValueError(
+            f"Histogram doesn't fit the data({max_val}). The max value is {max_val} but the histogram has {buckets} buckets with a width of"
+            f" {bucket_size}, the maximum value it can fit is {buckets * bucket_size}"
+        )
+
+    hs = [0] * buckets
+
+    for v in sorted_values:
+        bucket_index = int(v / bucket_size)
+        hs[bucket_index] += 1
+
+    return hs
+
+
+async def capture_traffic(container_name: str, duration_s: int) -> str:
+    cmd_rm = f"docker exec --privileged {container_name} rm /home/capture.pcap"
+    os.system(cmd_rm)
+
+    iface = "any"
+    capture_path = "/home/capture.pcap"
+
+    cmd = f"docker exec -d --privileged {container_name} tcpdump -i {iface} -U -w {capture_path}"
+    res = os.system(cmd)
+    if res != 0:
+        raise RuntimeError(f"Failed to launch tcpdump on {container_name}")
+
+    await asyncio.sleep(duration_s)
+
+    local_path = f"{container_name}.pcap"
+    subprocess.run([
+        "docker",
+        "cp",
+        container_name + ":" + "/home/capture.pcap",
+        local_path,
+    ])
+
+    cmd_rm = f"docker exec --privileged {container_name} pkill tcpdump"
+    os.system(cmd_rm)
+
+    return local_path
+
+
+# Render ASCII histogram drawing for visual inspection
+def draw_histogram(name: str, data: List[int], max_height=None):
+    if not data:
+        print(f"No data provided for {name}")
+        return
+
+    max_value = max(data)
+
+    if max_height is None:
+        max_height = max_value
+
+    scaled_data = [math.ceil((value / max_value) * max_height) for value in data]
+    for row in range(max_height, 0, -1):
+        line = ""
+        for value in scaled_data:
+            if value >= row:
+                line += "█"
+            else:
+                line += " "
+        line = "|" + line
+        print(line)
+
+    print(f"+{'-' * (len(data))}")
+    print(f"0{' ' * (len(data)-1)}{len(data)}")
+    print(f"^-Histogram of {name}")
+
+
+def generate_histogram_from_pcap(
+    pcap_path: str,
+    buckets: int,
+    bucket_width: int,
```

Comment:
I can answer that partially - I do not see any possibility of this value being higher than 1, since we can just simulate, but I can see a usage of `<1` bucket size. That would of course require seconds->milliseconds conversion as well.

Reply:
Refactored slighly that the call site doesn't pass the value of 1 but the function internally has a default of such, as a compromise